src/lib/datepicker/datepicker-toggle.ts

```diff
@@ -0,0 +1,121 @@
+import {
+    AfterContentInit,
+    Attribute,
+    ChangeDetectionStrategy,
+    ChangeDetectorRef,
+    Component,
+    ContentChild,
+    Directive,
+    Input,
+    OnChanges,
+    OnDestroy,
+    SimpleChanges,
+    ViewEncapsulation,
+    ViewChild
+} from '@angular/core';
+import { coerceBooleanProperty } from '@ptsecurity/cdk/coercion';
+import { McButton } from '@ptsecurity/mosaic/button';
+import { merge, of as observableOf, Subscription } from 'rxjs';
+
+import { McDatepicker } from './datepicker';
+import { McDatepickerIntl } from './datepicker-intl';
+
+
+/** Can be used to override the icon of a `mcDatepickerToggle`. */
+@Directive({
+    selector: '[mcDatepickerToggleIcon]'
+})
+export class McDatepickerToggleIcon {
+}
+
+
+@Component({
+    selector: 'mc-datepicker-toggle',
+    templateUrl: 'datepicker-toggle.html',
+    styleUrls: ['datepicker-toggle.css'],
+    host: {
+        class: 'mc-datepicker-toggle',
+        // Always set the tabindex to -1 so that it doesn't overlap with any custom tabindex the
+        // consumer may have provided, while still being able to receive focus.
+        '[attr.tabindex]': '-1',
+        '[class.mc-datepicker-toggle_active]': 'datepicker && datepicker.opened',
+        '[class.mc-warn]': 'datepicker && datepicker.color === "error"',
+        '(focus)': 'button.focus()'
+    },
+    exportAs: 'mcDatepickerToggle',
+    encapsulation: ViewEncapsulation.None,
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class McDatepickerToggle<D> implements AfterContentInit, OnChanges, OnDestroy {
+
+    /** Whether the toggle button is disabled. */
+    @Input()
+    get disabled(): boolean {
+        return this._disabled === undefined ? this.datepicker.disabled : !!this._disabled;
```

Comment:
это было взято из material, убрал

Reply:
подумал и вернул назад и поменял тип на boolean | null, т.к. disabled может быть не задан и это должно трактоваться как false

Reply:
вернул назад ок, тут специфичный случай, но `boolean | null` что тебе дает ?

Reply:
как раз то, что я описал, 3 состояния вместо 2х:
true - true
false - false
null\undefined etc - false

плюс отсутствие тупого кода, за который цепляется глаз

Reply:
> By default null and undefined are subtypes of all other types. That means you can assign null and undefined to something like number.

+ сеттер всегда установит boolean
```typescript
 set disabled(value: boolean) {
        this._disabled = coerceBooleanProperty(value);
}
```

в общем не выдумывай, null'a там не будет никогда.

Reply:
> In strict null checking mode, the null and undefined values are not in the domain of every type and are only assignable to themselves and any (the one exception being that undefined is also assignable to void). So, whereas T and T | undefined are considered synonymous in regular type checking mode (because undefined is considered a subtype of any T), they are different types in strict type checking mode, and only T | undefined permits undefined values. The same is true for the relationship of T to T | null

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html


У нас strictNullChecks включен

Reply:
т.е. ты хочешь сказать что _disabled теперь может быть null ?
покажи как ты это сделаешь ?

Reply:
когда @input пустой, но кто то уже читает из get disabled. В это время private _disabled пустой === undefined. Я заменил тип на `boolean | undefined`

Reply:
до вызова сеттера в _disabled  будет undefined

Reply:
повторяю вопросы:

> покажи как установить undefined в _disabled ?
> и ответь на вопрос зачем в данном случае для него сделаны get disabled и set disabled ?

и если ты не можешь на них ответить то убирай `| undefined`

Reply:
установить undefined в _disabled нет возможности, но сам _disabled undefined изначально, т.к. ничем не инициализирован
аксессоры сделаны для того, что бы обрабатывать такую то логику и хранить в _disabled правильные данные

сама строчка `this._disabled === undefined ` говорит что там может быть undefined. Спорить дальше не хочу, мне проще убрать :)

Reply:
убрал, теперь там `boolean`

Reply:
> но сам _disabled undefined изначально, т.к. ничем не инициализирован

да, и так для любой не инициализированной переменной любого типа, и в таком случае не имеет смысла делать : `string | undefined;` или `number | undefined;`.